DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp

```diff
@@ -0,0 +1,347 @@
+
+  return correct_result == kN;
+}
+
+int main(int argc, char *argv[]) {
+  auto begin_runtime = std::chrono::high_resolution_clock::now();
+
+  outfile.open("report.txt", std::ios_base::out);
+
+  std::vector<float> input_matrix(kN * kN);
+  std::vector<real> input_results(kN);
+
+  std::vector<queue> q;
+  for (const auto &p : platform::get_platforms()) {
+    if (p.get_info<info::platform::name>().find("Level-Zero") !=
+        std::string::npos) {
+      for (const auto &d : p.get_devices()) {
+        if (d.is_gpu() && d.get_info<info::device::name>().find("Intel") !=
+                              std::string::npos) {
+          q.push_back(queue(d));
+          std::cout << "--Found GPU: " << d.get_info<info::device::name>()
+                    << "\n";
+        }
+      }
+    }
+  }
+
+  if (q.size() < 2) {
+    std::cout << "Two GPUs are needed to run this code. Aborting\n";
+    return 0;
+  }
+
+  auto begin_matrix = std::chrono::high_resolution_clock::now();
+
+  GenerateMatrix(input_matrix, input_results);
+
+  buffer bufin_mat(input_matrix);
+  buffer bufin_res(input_results);
+
+  auto end_matrix = std::chrono::high_resolution_clock::now();
+  auto elapsed_matrix = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      end_matrix - begin_matrix);
+
+  std::cout << "\nMatrix generated, time elapsed: "
+            << elapsed_matrix.count() * 1e-9 << " seconds.\n";
+  outfile << "\nMatrix generated, time elapsed: "
+          << elapsed_matrix.count() * 1e-9 << " seconds.\n";
+
+  if (kN < 10) PrintMatrix(input_matrix, input_results);
+
+  auto begin_computations = std::chrono::high_resolution_clock::now();
+
+  std::vector<real> output_data(kN, 0);
+  std::vector<real> old_output_data(kN, 0);
+
+  for (int i = 0; i < kN; i++) output_data[i] = 0;
+
+  bool is_equal = false;
+  int sweeps = 0;
+
+  // The main functionality of the Jacobi Solver. Every iteration
+  // calculates new values until the difference between the values
+  // calculatedthis iteration and the one before is less than the error.
+  {
+    do {
+      buffer bufout_data(output_data);
+      buffer bufold_out_data(old_output_data);
+      for (int i = 0; i < kN; ++i) old_output_data[i] = output_data[i];
+      q[0].submit([&](handler &h) {
+            accessor D{bufout_data, h};
+            accessor OV{bufold_out_data, h};
+            accessor M{bufin_mat, h, read_only};
+            accessor R{bufin_res, h, read_only};
+            h.parallel_for(range<1>(kN / 2), [=](id<1> id) {
+              int i = id;
+              int j = kN * i;
+              int it = kN * i + i;
+
+              D[i] = R[i];
+              for (int z = 0; z < kN; ++z) {
+                if (z != i) D[i] = D[i] - (OV[z] * static_cast<real>(M[j]));
+                j = j + 1;
+              }
+              D[i] = D[i] / static_cast<real>(M[it]);
+            });
+          })
+          .wait();
+
+      q[1].submit([&](handler &h) {
+            accessor D{bufout_data, h};
```

Comment:
Same for D. OV, M, R.     This are not idiomatic like "id" or "i", "j", "k" (loop counters).